src/osp/tasks/top_utils.h

```diff
@@ -0,0 +1,183 @@
+{
+    entt::any &rData = topData[std::size_t(id)];
+    rData.emplace<T>(std::forward<ARGS_T>(args) ...);
+    return entt::any_cast<T&>(rData);
+}
+
+template <typename T, typename ANY_T>
+T& top_assign(ArrayView<entt::any> const topData, TopDataId id, ANY_T&& any)
+{
+    entt::any &rData = topData[std::size_t(id)];
+    rData = std::forward<ANY_T>(any);
+    return entt::any_cast<T&>(rData);
+}
+
+
+template <typename T>
+[[nodiscard]] T& top_get(ArrayView<entt::any> const topData, TopDataId const id)
+{
+    return entt::any_cast<T&>(topData[id]);
+}
+
+
+
+template<typename FUNCTOR_T>
+struct wrap_args_trait
+{
+    template<typename ... ARGS_T, std::size_t ... INDEX_T>
+    static constexpr void apply_void(ArrayView<entt::any> topData, [[maybe_unused]] std::integer_sequence<std::size_t, INDEX_T...> indices) noexcept
+    {
+        FUNCTOR_T{}(entt::any_cast<ARGS_T&>(topData[INDEX_T]) ...);
+    }
+
+    template<typename ... ARGS_T>
+    static TopTaskStatus wrapper_void([[maybe_unused]] WorkerContext& rCtx, ArrayView<entt::any> topData) noexcept
+    {
+        assert(topData.size() == sizeof...(ARGS_T));
+
+        apply_void<ARGS_T ...>(topData, std::make_integer_sequence<std::size_t, sizeof...(ARGS_T)>{});
+
+        return TopTaskStatus::Success;
+    }
+
+    template<typename ... ARGS_T, std::size_t ... INDEX_T>
+    static constexpr TopTaskStatus apply(ArrayView<entt::any> topData, [[maybe_unused]] std::integer_sequence<std::size_t, INDEX_T...> indices) noexcept
+    {
+        return FUNCTOR_T{}(entt::any_cast<ARGS_T&>(topData[INDEX_T]) ...);
+    }
+
+    template<typename ... ARGS_T>
+    static TopTaskStatus wrapper([[maybe_unused]] WorkerContext& rCtx, ArrayView<entt::any> topData) noexcept
+    {
+        assert(topData.size() == sizeof...(ARGS_T));
+
+        return apply<ARGS_T ...>(topData, std::make_integer_sequence<std::size_t, sizeof...(ARGS_T)>{});
+    }
+
+    template<typename ... ARGS_T>
+    static constexpr TopTaskFunc_t unpack([[maybe_unused]] TopTaskStatus(*func)(ARGS_T...))
+    {
+        // Parameter pack used to extract function arguments, func is discarded
+        return &wrapper<ARGS_T ...>;
+    }
+
+    template<typename ... ARGS_T>
+    static constexpr TopTaskFunc_t unpack([[maybe_unused]] void(*func)(ARGS_T...))
+    {
+        // Parameter pack used to extract function arguments, func is discarded
+        return &wrapper_void<ARGS_T ...>;
+    }
+};
+
+/**
+ * @brief Wrap a function with arbitrary arguments into a TopTaskFunc_t
+ *
+ * A regular TopTaskFunc_t accepts a ArrayView<entt::any> for passing data of
+ * arbitrary types. This needs to be manually casted.
+ *
+ * wrap_args creates a wrapper function that will automatically cast the
+ * entt::anys and call the underlying function.
+ *
+ * @param a[in]
+ *
+ * @return TopTaskFunc_t function pointer to wrapper
+ */
+template<typename FUNC_T>
+constexpr TopTaskFunc_t wrap_args(FUNC_T const&& funcArg)
```

Comment:
const&& doesn't exist.

What are you trying to do here?

Reply:
Assuming you're not referring to the doc, I put const there likely because funcArg isn't modified? This function only needs to get the type FUNC_T to be able to convert it to a function pointer.